spinalcordtoolbox/image.py

```diff
@@ -501,9 +517,17 @@ def save(self, path=None, dtype=None, verbose=1, mutable=False):
         hdr = self.hdr.copy() if self.hdr else None
         if hdr:
             hdr.set_data_shape(data.shape)
-            # Update dtype if provided (but not if based on SCT-specific values: 'minimize')
-            if (dtype is not None) and (dtype not in ['minimize', 'minimize_int']):
-                hdr.set_data_dtype(dtype)
+            # Update dtype to match that of the data array
+            dtype_data = data.dtype
+            dtype_header = hdr.get_data_dtype()
+            if dtype_data != dtype_header:
+                logger.warning(f"DTYPEWARN: Image header specifies datatype '{dtype_header}', but array is of type"
+                               f"type '{dtype_data}'. Header metadata will be overwritten to use '{dtype_data}'.")
+                if dtype_data == 'bool':
+                    # Sometimes we create boolean arrays, apparently! But, using 'bool' for nibabel headers triggers:
+                    # `nibabel.spatialimages.HeaderDataError: data dtype "bool" not supported`, so use `int16` instead
+                    dtype_data = np.int16
+                hdr.set_data_dtype(dtype_data)
```

Comment:
The test suite [reveals](https://github.com/spinalcordtoolbox/spinalcordtoolbox/actions/runs/3448711626/jobs/5755958619#step:5:834) that in several places, we create boolean arrays, then attempt to save them as image files.

Now, `nibabel` can handle the actual boolean array just fine when saving. But, if you try and call `hdr.set_data_dtype('bool')`, then you'll get an error, because `bool` isn't a valid header dtype value. 

So, just to mitigate some CI errors, I've substituted `int16` in, so that the image gets saved as an `int16` image. But, I  think the actual long-term solution deserves a little more discussion:

- What datatype should boolean arrays be cast to?
- When/where should this casting happen? 
- Should we avoid creating images from boolean arrays at all?

Reply:
For boolean arrays, I suspect `np.uint8` is the natural header dtype to use, rather than `np.int16`, assuming we want to use integer types at all.

Reply:
> For boolean arrays, I suspect `np.uint8` is the natural header dtype to use, rather than `np.int16`, assuming we want to use integer types at all.

`np.uint8` sounds good. It is used for boolean data by the `nilearn`; see [here](https://github.com/nilearn/nilearn/pull/3286)

Reply:
`np.uint8` was my gut reaction too, but then @mguaypaq brought up an interesting point in a Slack message:

> * unsigned types are pretty terrible. For example:
>    ```python
>    >>> import numpy as np
>    >>> x = np.zeros(1, dtype=np.uint8); x
>    array([0], dtype=uint8)
>    >>> x - 1
>    array([255], dtype=uint8)
>     ```

Given that our binary arrays tend to be masks, and given that it's possible for masks to be used in array arithmetic (e.g. https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3843), I was a little hesitant to use an unsigned type? 

That said, `uint8` _is_ used quite heavily throughout the SCT codebase for binary images. So, it's not like using `uint8` would be out of place. Perhaps we should stick with `uint8`, then discuss SCT's overall handling of `uint8` images in a separate issue?

Reply:
I'm thinking more and more about this possible pattern:
- For masks and labels that we generate, store them as `uint8` (assuming we never see very large labels, i.e., bigger than 255?)
- When doing any arithmetic at all, convert all inputs to `float64`, and store the results as `float64`.

At this point my reasons for this are more intuitions than anything very solid, but do you have any immediate thoughts?

Reply:
(There's maybe the question of whether some operations, like `min` and `max`, count as arithmetic, but I think that can be resolved later.)

Reply:
I'll note that, despite its bad reputation, the `float64` type can handle arithmetic with integer values just fine. In particular, doing additions, subtractions and multiplications with any integers that would fit inside 32 bits without overflow gives the same result in `float64`, without any loss of precision. Similarly, divisions where the result is an integer (so, no fractional part) also have no loss of precision.

Reply:
Ah, except that `sct_image -type`, which is definitely user-facing, lets the user choose the storage type from `('uint8', 'int16', 'int32', 'float32', 'complex64', 'float64', 'int8', 'uint16', 'uint32', 'int64', 'uint64')`. The internal methods also accept `minimize` and `minimize_int`, but I can't find any use of them in SCT (so maybe they can be dropped?).

Reply:
I've typed up a response, but... I'm wondering if it might be better to move this discussion into a separate issue, given that the task of standardizing datatypes is much greater in scope than this PR accounts for.

(For the record, though, I do agree with the general notion of using `uint8` for labels/masks, and `float64` for arithmetic.)